src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageProcessingConnector.cs

```diff
@@ -0,0 +1,135 @@
+
+            var deduplicationEntry = await outboxStorage.Get(messageId, new OutboxStorageOptions(context)).ConfigureAwait(false);
+            var pendingTransportOperations = new PendingTransportOperations();
+
+            if (deduplicationEntry == null)
+            {
+                physicalMessageContext.Set(pendingTransportOperations);
+
+                await next(physicalMessageContext).ConfigureAwait(false);
+
+                var outboxMessage = new OutboxMessage(messageId, ConvertToOutboxOperations(pendingTransportOperations.Operations).ToList());
+
+                await outboxStorage.Store(outboxMessage, new OutboxStorageOptions(context)).ConfigureAwait(false);
+            }
+            else
+            {
+                ConvertToPendingOperations(deduplicationEntry, pendingTransportOperations);
+            }
+
+            if (pendingTransportOperations.Operations.Any())
+            {
+                var batchDispatchContext = new BatchDispatchContext(pendingTransportOperations.Operations, physicalMessageContext);
+
+                await batchDispatchPipeline.Invoke(batchDispatchContext).ConfigureAwait(false);
+            }
+
+            await outboxStorage.SetAsDispatched(messageId, new OutboxStorageOptions(context)).ConfigureAwait(false);
+        }
+
+        void ConvertToPendingOperations(OutboxMessage deduplicationEntry, PendingTransportOperations pendingTransportOperations)
+        {
+            foreach (var operation in deduplicationEntry.TransportOperations)
+            {
+                var options = new DispatchOptions(DeserializeRoutingStrategy(operation.Options),
+                    DispatchConsistency.Isolated,
+                    DeserializeConstraints(operation.Options));
+
+                var message = new OutgoingMessage(operation.MessageId, operation.Headers, operation.Body);
+
+                pendingTransportOperations.Add(new TransportOperation(message, options));
+            }
+        }
+
+        IEnumerable<Outbox.TransportOperation> ConvertToOutboxOperations(IEnumerable<TransportOperation> operations)
+        {
+            foreach (var operation in operations)
+            {
+                var options = new Dictionary<string, string>();
+
+                operation.DispatchOptions.DeliveryConstraints.ToList().ForEach(c => c.Serialize(options));
+                operation.DispatchOptions.RoutingStrategy.Serialize(options);
+
+                yield return new Outbox.TransportOperation(operation.Message.MessageId,
+                    options, operation.Message.Body, operation.Message.Headers);
+            }
+        }
+
+        public IEnumerable<DeliveryConstraint> DeserializeConstraints(Dictionary<string, string> options)
+        {
+            if (options.ContainsKey("NonDurable"))
+            {
+                yield return new NonDurableDelivery();
+            }
+
+            string deliverAt;
+            if (options.TryGetValue("DeliverAt", out deliverAt))
```

Comment:
Not sure I like the fact that deserialization and serialization is done in two separate places. I know that the problem with deserialization is lack of type and we need this factory method but maybe it would be better (for consistence) to also move serialization here to this class?